pilot/platform/cloudfoundry/controller_test.go

```diff
@@ -0,0 +1,192 @@
+				},
+				"process-guid-b.cfapps.internal": &api.BackendSet{
+					Backends: []*api.Backend{
+						{
+							Address: "10.0.50.4",
+							Port:    61009,
+						},
+						{
+							Address: "10.0.60.2",
+							Port:    61001,
+						},
+					},
+				},
+			},
+		}
+		client.RoutesOutput.Ret1 <- nil
+		controller = &cloudfoundry.Controller{
+			Client: client,
+			Ticker: ticker,
+		}
+	})
+
+	It("calls all the instance handlers when the cache is invalidated", func() {
+		ih1, ih2 := new(fakeInstanceHandler), new(fakeInstanceHandler)
+		sh1, sh2 := new(fakeServiceHandler), new(fakeServiceHandler)
+
+		controller.AppendInstanceHandler(ih1.Do)
+		controller.AppendInstanceHandler(ih2.Do)
+		controller.AppendServiceHandler(sh1.Do)
+		controller.AppendServiceHandler(sh2.Do)
+
+		stop := make(chan struct{})
+		defer close(stop)
+		go controller.Run(stop)
+
+		By("checking no handlers are called before the ticker fires")
+		Consistently(client.RoutesCalled, "100ms").ShouldNot(Receive())
+		Consistently(ih1.callCount, "100ms").Should(Equal(0))
+		Consistently(ih2.callCount, "100ms").Should(Equal(0))
+		Consistently(sh1.callCount, "100ms").Should(Equal(0))
+		Consistently(sh2.callCount, "100ms").Should(Equal(0))
+
+		By("checking that all handlers are called after the first ticker fires")
+		ticker <- time.Time{}
+		Eventually(client.RoutesCalled).Should(Receive())
+		Eventually(ih1.callCount).Should(Equal(1))
+		Eventually(ih2.callCount).Should(Equal(1))
+		Eventually(sh1.callCount).Should(Equal(1))
+		Eventually(sh2.callCount).Should(Equal(1))
+
+		By("checking that no handlers are called if the cached data is still valid")
+		ticker <- time.Time{}
+		Eventually(client.RoutesCalled).Should(Receive())
+		Consistently(ih1.callCount, "100ms").Should(Equal(1))
+		Consistently(ih2.callCount, "100ms").Should(Equal(1))
+		Consistently(sh1.callCount, "100ms").Should(Equal(1))
+		Consistently(sh2.callCount, "100ms").Should(Equal(1))
+
+		By("checking that all handlers are called again when the cache is invalidated")
+		client.RoutesOutput.Ret0 <- &api.RoutesResponse{
+			Backends: map[string]*api.BackendSet{
+				"other-process-guid-a.cfapps.internal": &api.BackendSet{
+					Backends: []*api.Backend{
+						{
+							Address: "10.10.2.6",
+							Port:    61006,
+						},
+						{
+							Address: "10.0.41.3",
+							Port:    61009,
+						},
+					},
+				},
+				"process-guid-b.cfapps.internal": &api.BackendSet{
+					Backends: []*api.Backend{
+						{
+							Address: "10.0.50.4",
+							Port:    61009,
+						},
+						{
+							Address: "10.0.60.2",
+							Port:    61001,
+						},
+					},
+				},
+			},
+		}
+		client.RoutesOutput.Ret1 <- nil
+		ticker <- time.Time{}
+		Eventually(client.RoutesCalled).Should(Receive())
+		Eventually(ih1.callCount).Should(Equal(2))
+		Eventually(ih2.callCount).Should(Equal(2))
+		Eventually(sh1.callCount).Should(Equal(2))
+		Eventually(sh2.callCount).Should(Equal(2))
+	})
+})
+
+type fakeTicker chan time.Time
```

Comment:
Are there other examples of doing this sort of thing within Istio/golang? It would seem generally beneficial to have a common mockable ticker that we can all share.

Reply:
This is a pattern that we use commonly in cloudfoundry specifically in routing team when you want to avoid `time.Sleeps` in your test and want to take care of ticking yourself. We can certainly extract this into a separate package and re-use in other places if there is a need for this in other areas of istio.